diff -uNr libproxy-0.4.10/libproxy/test/get-pac-test.cpp libproxy-0.4.10.new/libproxy/test/get-pac-test.cpp
--- libproxy-0.4.10/libproxy/test/get-pac-test.cpp	2012-10-15 22:32:00.000000000 +0200
+++ libproxy-0.4.10.new/libproxy/test/get-pac-test.cpp	2012-10-29 18:38:57.000000000 +0100
@@ -3,7 +3,11 @@
 #include <sstream>
 #include <string>
 
+
+#include <sys/types.h>
+
 #include <sys/socket.h>
+#include <sys/select.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
 #include <unistd.h>
@@ -32,7 +36,7 @@
 			int i = 1;
 
 			addr.sin_family = AF_INET;
-			inet_aton("127.0.0.1", &addr.sin_addr);
+			// inet_aton("127.0.0.1", &addr.sin_addr);
 			addr.sin_port = ntohs(m_port);
 
 			if (m_sock != -1)
